### PR TITLE
ETC-9: Add ToEtcExpr moduless

### PR DESCRIPTION
**Pull request by shonfeder**

### CODE REVIEW

**Review by konnov — APPROVED** [review]
Looks good!

**Comment by shonfeder** [comment]
Still WIP, I should have marked it as such! But thank you for the quick review :)

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+              s"Parser error in Typing!EmptySeq($elemTypeText): ${e.msg}"
+            )
+        }
+
+      //******************************************** GENERAL OPERATORS **************************************************
+      case OperEx(op, args @ _*) if op == TlaOper.eq || op == TlaOper.ne =>
+        // x = y, x /= y
+        val a = varPool.fresh
+        val opsig = OperT1(Seq(a, a), BoolT1())
+        mkApp(ex.ID, opsig, args)
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:166]: @konnov why don't we use an `ExactRef` here?

**Comment by shonfeder** [comment]
Removed WIP prefix (which I should have done on Friday).

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 2]
```diff
@@ -0,0 +1,858 @@
...
+  *
+  * @author Igor Konnov
+  */
+class ToEtcExpr(varPool: TypeVarPool) extends EtcBuilder {
+  private val type1Parser = DefaultType1Parser
+
+  /**
+    * The counter that we use to produce fresh variables
+    */
+  private var nextVarNum = 0
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:24]: I think this variable is left after the previous version. The current version is using `varPool`, so `nextVarNum` can be safely deleted.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 3]
```diff
@@ -0,0 +1,858 @@
...
+
+  /**
+    * Get the number just above the maximum variable number
+    * @return an upper bound on the variable numbers
+    */
+  def varNumUpperBound: Int = nextVarNum
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:30]: Same here. This must be the leftover after the old version, which can be safely removed.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 4]
```diff
@@ -0,0 +1,858 @@
...
+      throw new IllegalArgumentException(s"Expected key-set pair, found ${orphan}")
+
+    // since we group by 2 below, this case should be unreachable
+    case moreThanTwo =>
+      throw new IllegalArgumentException(
+        s"Reached impossible state in validateRecSetPair: ${moreThanTwo}"
+      )
+  }
+
+  // TODO: a long string of translation rules. Can we decompose it?
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:144]: shall we remove this TODO?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 5]
```diff
@@ -0,0 +1,858 @@
...
+      case LetInEx(body, declarations @ _*) =>
+        declarations.foldRight(this(body)) {
+          case (decl, term) => this(decl, term)
+        }
+
+      //******************************************** BOOLEANS **************************************************
+      case OperEx(op, args @ _*)
+          if op == TlaBoolOper.and || op == TlaBoolOper.or || op == TlaBoolOper.equiv || op == TlaBoolOper.implies =>
+        // A /\ B, A \/ B, A <=> B, A => B
+        val nBools = List.fill(args.length)(BoolT1())
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:250]: Technically, `equiv` and `implies` admit only 2 arguments. Perhaps the translator should complain immediately, if this is not the case.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 6]
```diff
@@ -0,0 +1,858 @@
...
+        }
+
+      case _ =>
+        throw new TypingInputException(
+          s"""Error in ${decl.name}: Expected /\ AssumeType(varName, "typeString") /\ ..."""
+        )
+    }
+  }
+
+  private val typeOfLiteralExpr: TlaValue => TlaType1 = {
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:113]: nice!

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 7]
```diff
@@ -0,0 +1,858 @@
...
+        val principal = OperT1(Seq(OperT1(typeVars, BoolT1())), SetT1(funFrom))
+
+        // map implicitly introduces a lambda abstraction: λ x ∈ S, y ∈ T. P
+        val lambda = mkAbs(BlameRef(ex.ID), this(pred), bindings: _*)
+        // the resulting expression is ((((a, b) => Bool) => Set(<<a, b>>)) (λ x ∈ S, y ∈ T. P)
+        mkApp(ref, Seq(principal), lambda)
+
+      case OperEx(TlaSetOper.map, mapExpr, args @ _*) =>
+        // { x \in S, y \in T: e }
+        val (vars, sets) = args.zipWithIndex.partition(_._2 % 2 == 0)
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:369]: Shall we use `TlaOper.deinterleave` now?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 8]
```diff
@@ -0,0 +1,858 @@
...
+        val b1 = varPool.fresh
+        // introduce a sequence of a, b, a, b, ... (as many as there are accessors)
+        val aAndBs = List.fill(accessors.length)(List(a1, b1)).flatten
+        val funType = OperT1(FunT1(a1, b1) +: aAndBs, FunT1(a1, b1))
+
+        val typeVars = varPool.fresh(accessors.length) // introduce type variables a, b, c, ...
+        val accessorsWithTypeVars = accessors.zip(typeVars)
+
+        // The following values are related by mutual exclusion, so we make
+        // them lazy to avoid unnecesary computations
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:520]: good idea!

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala [hunk 9]
```diff
@@ -0,0 +1,858 @@
...
+
+      case OperEx(TlaActionOper.stutter, args @ _*) =>
+        // Bool, a, b, c => Bool
+        val opsig = OperT1(BoolT1() +: varPool.fresh(args.length - 1), BoolT1())
+        mkExRefApp(opsig, args)
+
+      case OperEx(TlaActionOper.nostutter, args @ _*) =>
+        // Bool, a, b, c => Bool
+        val opsig = OperT1(BoolT1() +: varPool.fresh(args.length - 1), BoolT1())
+        mkExRefApp(opsig, args)
```
Comment (konnov) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:649]: we could merge these two into one case

**Review by konnov — APPROVED** [review]
I like it a lot. Nice improvements!

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala (hunk 9 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:649]: Oh yeah, I missed that! Nice catch.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala (hunk 7 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:369]: Good idea!

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala (hunk 5 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala:250]: Yeah, let's do catch that. I'll separate those out.

**Pull request merged**